lib/spack/spack/spec.py

```diff
@@ -2979,8 +2979,8 @@ def normalize(self, force=False, tests=False, user_spec_deps=None):
                 "Attempting to normalize anonymous spec")
 
         # Set _normal and _concrete to False when forced
-        if force:
-            self._mark_concrete(False)
+        if force and not self._concrete:
+            self._normal = False
```

Comment:
Side note: without this the old concretizer will "flatten" concrete already-installed specs but doesn't necessarily have the info to reassemble them, so we avoid disconnecting concrete sub-specs.